docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,235 @@
+| `isnever()` | `boolean` | returns true if self is a runtime representation of the built-in type `never` |
+| `isany()` | `boolean` | returns true if self is a runtime representation of the built-in type `any` |
+| `isnegation()` | `boolean` | returns true if self is a runtime representation of a `Negation` |
+| `isboolean()` | `boolean` | returns true if self is a runtime representation of the built-in type`boolean` |
+| `isnumber()` | `boolean` | returns true if self is a runtime representation of the built-in type `number` |
+| `isstring()` | `boolean` | returns true if self is a runtime representation of the built-in type `string` |
+| `isstringsingleton()` | `boolean` | returns true if self is a runtime representation of a string singleton |
+| `isbooleansingleton()` | `boolean` | returns true if self is a runtime representation of a boolean singleton |
+| `isunion()` | `boolean` | returns true if self is a runtime representation of the union type |
+| `isintersection()` | `boolean` | returns true if self is a runtime representation of the intersection type |
+| `istable()` | `boolean` | returns true if self is a runtime representation of a `table` type |
+| `ismetatable()` | `boolean` | returns true if self is a runtime representation of `Metatable` |
+| `isfunction()` | `boolean` | returns true if self is a runtime representation of a `function` type |
+| `isclass()` | `boolean` | returns true if self is a runtime representation of a `class` type |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is a subtype or equal to arg in the type hierarchy |
+| `conformsto(arg: typelib)` | `boolean` | returns true if self is equal to arg in the type hierarchy |
+
+#### Negation
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `gettype()` | `typelib` | returns the runtime representation of the type being negated |
+
+#### String
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getmetatable()` | `typelib` | returns the runtime representation of `Metatable` |
+
+#### StringSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `string` | returns a string singleton |
+
+#### BooleanSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `boolean` | returns either `true` or `false` |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
+| `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |
```

Comment:
these functions take singular properties, so they should be named `addprop` and `delprop`

Reply:
We could use `__newindex` for this.

```luau
t["myprop"] = mytype -- adds the prop
t["myprop"] = nil -- deletes the prop
```

Reply:
Yeah, that would be considerably nicer.

Reply:
That works, in my mind, if we actually make `{ [ "foo" | "bar" | "baz"]: string }` equivalent to `{ foo: string, bar: string, baz: string }`, but I don't think it's so bad if indexers are an explicit thing.